h-refinement

wthrowe opened the pull request:
This does not include methods of actually making refined domains, but includes everything you would need to evolve on such a domain if you had one.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Add function to determine size of mortar

wthrowe committed:
Add insert_element array helper

wthrowe committed:
Handle h-refinement in MortarHelpers

wthrowe committed:
Add mortar sizes to DataBox

wthrowe committed:
Support h-refinement in ABFGTS

wthrowe committed:
Support h-refinement in SendDataForFluxes

kidder commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.cpp:
```diff
@@ -21,11 +26,43 @@ Mesh<Dim> mortar_mesh(const Mesh<Dim>& face_mesh1,
+    const size_t dimension, const OrientationMap<Dim>& orientation) noexcept {
+  const auto self_segments =
+      all_but_specified_element_of(self.segment_ids(), dimension);
+  const auto neighbor_segments = all_but_specified_element_of(
```
check that it is the inverse map that should be called

wthrowe replied:
It is being passed a neighbor quantity, so the inverse map is correct.

kidder commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.hpp:
```diff
@@ -32,6 +37,13 @@ template <size_t Dim>
                       const Mesh<Dim>& face_mesh2) noexcept;
 
+/// \ingroup DiscontinuousGalerkinGroup
+/// Determine the size of the mortar for communicating with a neighbor.
```
make clearer that size is full vs half and refers to the view from yourself (and therefore does not need to agree with what the neighbor will determine)

kidder commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_MortarHelpers.cpp:
```diff
@@ -43,6 +51,88 @@ SPECTRE_TEST_CASE("Unit.DG.MortarHelpers.mortar_mesh",
+  // Check the root segment to make sure the code doesn't try to get
+  // its parent.
+  CHECK(dg::mortar_size(ElementId<2>(0, {{{0, 0}, {0, 0}}}),
+                        ElementId<2>(0, {{{0, 0}, {0, 0}}}), 1, {}) ==
```
different block_ids for root elements

kidder commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_MortarHelpers.cpp:
```diff
@@ -43,6 +51,88 @@ SPECTRE_TEST_CASE("Unit.DG.MortarHelpers.mortar_mesh",
+  }
+
+  // Check an orientation case
+  CHECK(dg::mortar_size(ElementId<3>(1, {{{0, 0}, {3, 2}, {7, 5}}}),
```
comment that perp refinement levels are not intentionally within one as this function doesn't care

nilsdeppe replied:
Did you add this comment and I just can't find it?

wthrowe replied:
I noticed the same thing was happening above so I put the comment there.

kidder commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Tags.hpp:
```diff
@@ -35,6 +37,15 @@ struct Mortars : db::PrefixTag, db::SimpleTag {
+/// \ingroup DataBoxTagsGroup
+/// \ingroup DiscontinuousGalerkinGroup
+/// Size of a mortar, relative to an element face.
+template <size_t Dim>
```
same comment about size

kidder reviewed (commented):
you can squash

nilsdeppe commented on src/Utilities/StdArrayHelpers.hpp:
```diff
@@ -117,6 +117,33 @@ inline std::array<T, Dim - 1> all_but_specified_element_of(
+}
+
+template <size_t ElementToAdd, typename T, size_t Dim>
+inline std::array<T, Dim + 1> insert_element(std::array<T, Dim> a,
```
Why add this one if it's not constexpr?

wthrowe replied:
Symmetry with `all_but_specified_element_of`.  I can remove it if you prefer.  (I can remove the `all_but_specified_element_of` one too if you'd like.)

nilsdeppe replied:
Offline discussion: We will remove them since the main advantage of the template parameter would be constexpr. If we need a constexpr version we can just write it then :)

nilsdeppe reviewed (commented):
I'll deal with the rest after p-refinement is in. I'm hopeful this PR could be split into two or three since it's quite long...

wthrowe reviewed (commented):
This is actually smaller than the p-refinement PR, it just looks huge because it is currently on top of that one.

wthrowe committed:
Make all_but_specified_element_of param runtime

The only compile-time thing it was used for was a static_assert, and
it is useful to be able to pass a runtime value.

wthrowe commented:
I squashed the latest changes (first two commits) because it was easier to redo one commit than make the changes from the finished state.

nilsvu commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Tags.hpp:
```diff
@@ -35,6 +37,16 @@ struct Mortars : db::PrefixTag, db::SimpleTag {
+template <size_t Dim>
+struct MortarSize : db::SimpleTag {
+  static std::string name() noexcept { return "MortarSize"; }
+  using type = std::array<Spectral::MortarSize, Dim>;
```
I think the `MortarSize` type should be in `dg` instead of `Spectral`, since also this tag (rightly) is. I suggest to move `Projection.*pp` to `NumericalAlgorithms/DiscontinuousGalerkin`

nilsdeppe replied:
I don't really see the argument here because I view projection as a spectral operation: it is a method of best-representing a function in one space in a different space. That alone is not dG specific. Having the tag inside dG I view as an application of those functions.

nilsvu replied:
I agree with the projection operation being in `Spectral`. But it bothers me a little that `MortarSize` is in `Spectral`, since that's more of a `Domain` or `dg` thing..

wthrowe replied:
Now that you point it out MortarSize is not a great name.  Not coming up with anything better on the spot.

nilsdeppe replied:
Okay, I'm merging, we can rename later when we come up with a better name. Added an issue in #870, need to keep moving :)

nilsdeppe commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_MortarHelpers.cpp:
```diff
@@ -141,107 +141,187 @@ namespace {
-          func(face_coords));
     }
   }
   // Check 2D
```
Am I not seeing a 3D check or did we decide to not check that?

wthrowe replied:
We decided not to check that.

nilsdeppe replied:
Sounds good!

nilsvu commented on src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.cpp:
```diff
@@ -21,11 +26,43 @@ Mesh<Dim> mortar_mesh(const Mesh<Dim>& face_mesh1,
           Spectral::Quadrature::GaussLobatto};
```
This function assumes that both meshes are on LGL points, correct? I suppose that's fine for now, but it would be good to add a comment about that. Also for the h-refined case where the face that covers the full mortar has less points than the larger face that is only half covered by the mortar, I think we talked about using Gauss points since both sides need to do an interpolation anyway.

BTW don't you have to adjust this function for h-refinement anyway, e.g. add a factor of 1/2 into the `std::max` call above if a face is larger than the mortar, or am I getting something wrong?

nilsdeppe replied:
1. An `ASSERT` would be safer than a comment, if anything.

2. You are indeed getting something wrong. Just because the mesh is half the size doesn't mean you can represent polynomials 2 * degree larger. This code is correct.

wthrowe replied:
An ASSERT or similar would be reasonable here, but this code isn't part of this PR; it's just at the edge of the diff context.  Make an issue or put in another PR.

nilsdeppe replied:
#869

nilsvu replied:
👍

How about using Gauss points for when both elements have to interpolate anyway, which is when the larger element has a higher resolution than the smaller element (I think)?
- Do we want to do that?
- Will this mess up anything in the current code?
- Should we change it in this PR or open an issue?

wthrowe replied:
> Do we want to do that?

We should certainly try it.

> Will this mess up anything in the current code?

I don't think it will, but I haven't tested.

> Should we change it in this PR or open an issue?

Open an issue.  We have a looming deadline and I don't want to spend time right now swapping code that is known to work out for something I'm not sure will work.

(Your conditions on when double interpolation happens are correct.)

nilsdeppe replied:
Yes, we need to move on. Filing an issue is the best way to get this changed in the future, but as @wthrowe said, we have a deadline and we can't spend time on anything not on the critical path.

nilsvu replied:
#871

nilsdeppe commented:
@wthrowe I'm happy for you to squash any changes you feel should be made, none of my comments are blocking :)

nilsdeppe commented:
@wthrowe Need to clean up IWYU and clang-tidy, then this can be merged.

nilsdeppe commented:
macOS build timed out when backing up cache. Merging anyway.
